build/postgresql/patches-14/dism.patch

```diff
@@ -0,0 +1,89 @@
+@@ -141,16 +141,16 @@
+ #autovacuum_work_mem = -1		# min 1MB, or -1 to use maintenance_work_mem
+ #logical_decoding_work_mem = 64MB	# min 64kB
+ #max_stack_depth = 2MB			# min 100kB
+-#shared_memory_type = mmap		# the default is the first option
++#shared_memory_type = sysv		# the default is the first option
+ 					# supported by the operating system:
+-					#   mmap
+ 					#   sysv
++					#   mmap
+ 					#   windows
+ 					# (change requires restart)
+-#dynamic_shared_memory_type = posix	# the default is the first option
++#dynamic_shared_memory_type = sysv	# the default is the first option
+ 					# supported by the operating system:
+-					#   posix
+ 					#   sysv
++					#   posix
+ 					#   windows
+ 					#   mmap
+ 					# (change requires restart)
+diff -wpruN '--exclude=*.orig' a~/src/include/portability/mem.h a/src/include/portability/mem.h
+--- a~/src/include/portability/mem.h	1970-01-01 00:00:00
++++ a/src/include/portability/mem.h	1970-01-01 00:00:00
+@@ -14,11 +14,15 @@
+ 
+ #define IPCProtection	(0600)	/* access/modify by user only */
+ 
++#ifdef SHM_PAGEABLE			/* use dynamic intimate shared memory on Solaris */
++#define PG_SHMAT_FLAGS			SHM_PAGEABLE
++#else
+ #ifdef SHM_SHARE_MMU			/* use intimate shared memory on Solaris */
+ #define PG_SHMAT_FLAGS			SHM_SHARE_MMU
+ #else
+ #define PG_SHMAT_FLAGS			0
+ #endif
++#endif
+ 
+ /* Linux prefers MAP_ANONYMOUS, but the flag is called MAP_ANON on other systems. */
+ #ifndef MAP_ANONYMOUS
+diff -wpruN '--exclude=*.orig' a~/src/include/storage/dsm_impl.h a/src/include/storage/dsm_impl.h
+--- a~/src/include/storage/dsm_impl.h	1970-01-01 00:00:00
++++ a/src/include/storage/dsm_impl.h	1970-01-01 00:00:00
+@@ -19,7 +19,14 @@
+ #define DSM_IMPL_WINDOWS		3
+ #define DSM_IMPL_MMAP			4
+ 
++#ifdef HAVE_SYS_SHM_H
+ /*
++ * For SHM_SHARE_MMU.
++ */
++#include <sys/shm.h>
++#endif
++
++/*
+  * Determine which dynamic shared memory implementations will be supported
+  * on this platform, and which one will be the default.
+  */
+@@ -34,7 +41,12 @@
+ #define USE_DSM_SYSV
+ #ifndef DEFAULT_DYNAMIC_SHARED_MEMORY_TYPE
+ #define DEFAULT_DYNAMIC_SHARED_MEMORY_TYPE		DSM_IMPL_SYSV
++#else
++#ifdef SHM_SHARE_MMU
++#undef DEFAULT_DYNAMIC_SHARED_MEMORY_TYPE
++#define DEFAULT_DYNAMIC_SHARED_MEMORY_TYPE		DSM_IMPL_SYSV
+ #endif
++#endif
+ #define USE_DSM_MMAP
+ #endif
+ 
+diff -wpruN '--exclude=*.orig' a~/src/include/storage/pg_shmem.h a/src/include/storage/pg_shmem.h
+--- a~/src/include/storage/pg_shmem.h	1970-01-01 00:00:00
++++ a/src/include/storage/pg_shmem.h	1970-01-01 00:00:00
+@@ -70,9 +70,9 @@
+ #endif
+ extern void *UsedShmemSegAddr;
+ 
+-#if !defined(WIN32) && !defined(EXEC_BACKEND)
++#if !defined(WIN32) && !defined(EXEC_BACKEND) && !defined(SHM_SHARE_MMU)
+ #define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_MMAP
+-#elif !defined(WIN32)
++#elif !defined(WIN32) || defined(SHM_SHARE_MMU)
+ #define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_SYSV
+ #else
+ #define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_WINDOWS
```

Comment:
It took me a minute to parse this, it might be better to patch it like:

```diff
@@ -70,7 +70,9 @@
 #endif
 extern void *UsedShmemSegAddr;

+#if defined(SHM_SHARE_MMU)
+#define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_SYSV
-#if !defined(WIN32) && !defined(EXEC_BACKEND)
+#elif !defined(WIN32) && !defined(EXEC_BACKEND)
 #define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_MMAP
 #elif !defined(WIN32)
 #define DEFAULT_SHARED_MEMORY_TYPE SHMEM_TYPE_SYSV
```

Happy either way, it's obviously doing the right thing.

Reply:
Yeah I already did the other for dynamic shared memory to be somewhat better readable. The patch however looks a bit different then what you pasted :-) diff does the grouping just a bit differently it seems. I tested with -xx on all 3 build scripts and the new dism.patch still applies to all of them now.